### ISSUE TITLE
[Discussion] Solutions for mis-matching PIDs in Windows

### ISSUE THREAD

**Issue by czgdp1807** [ray-project/ray#18951]
From @pcmoritz's ideas I came up with following approaches to deal with [the following inconsistency](https://github.com/ray-project/ray/issues/15970#issuecomment-927136860),

> Inconsistency - The output of `getpid` in [this line](https://github.com/ray-project/ray/blob/65fa740c3b8f27803715521524c4d2245f3b0517/src/ray/core_worker/core_worker.cc#L146) is different from `pi.dwProcessId`([here](https://github.com/ray-project/ray/blob/65fa740c3b8f27803715521524c4d2245f3b0517/src/ray/util/process.cc#L152)) for the same process. `pid = pi.dwProcessId` is under `#ifdef _WIN32` and `#endif`. So, may be this is difference in PIDs is the cause of whole problem in windows? 
> The above inconsistency only happens in virtual environment. In conda environment both the things are same and probably therefore the code doesn't hang there.  

**Discard PIDs in favour of UIDs**

In this approach, a process launched by Ray will be having a UID (sort of unique identity with respect to ray) instead of a PID assigned by the underlying OS. UID will be generated (how we do this is not of great concern here) inside Ray and passed as a command line argument [here](https://github.com/ray-project/ray/blob/96054953ccffe0281a29062942a4f57f6a1403df/src/ray/util/process.cc#L124-L126). We will replace `pid_t pid` member with `uid_t uid`(`uid_t` will be `int` in Windows and `int64_t` in other systems) in `ProcessFD` class (see below).

https://github.com/ray-project/ray/blob/96054953ccffe0281a29062942a4f57f6a1403df/src/ray/util/process.cc#L84-L85

In addition to the above, `pid`, `worker_shim_pid` arguments will be replaced by a single argument `uid` in `WorkerPool::RegisterWorker`(see https://github.com/ray-project/ray/blob/96054953ccffe0281a29062942a4f57f6a1403df/src/ray/raylet/worker_pool.h#L213-L215). There will be some rippling effects due to this and changes will be made to accommodate them. After all the above stuff, no matter how many wrapper processes a worker process has, it will always share its `uid` (assigned to it via command line arguments during its startup) which will always be there in `state.starting_worker_processes`.

In case we want to retain PIDs along side UIDs then I have thought of the following.

**Link PIDs with UIDs**

First create a new `unordered_map<uid_t uid, std::vector<pid_t>> uid2pids`. When starting the process, pass the generated UID to the worker process via command line arguments as we do in the above approach. Additionally, after https://github.com/ray-project/ray/blob/96054953ccffe0281a29062942a4f57f6a1403df/src/ray/util/process.cc#L155 do, `uid2pids[uid].push_back(pid)`. When sending request for registration in `WorkerPool::RegisterWorker`, we will send `uid_t uid` argument as well. Using this `uid`, we can find in `uid2pids` map that which `pid` is the actual one and then use it to perform `state.starting_worker_process.count(shim_process) == 0` (see, https://github.com/ray-project/ray/blob/a0a02f498222e5e2396298415262fe6b744bb1f1/src/ray/raylet/worker_pool.cc#L510)

The disadvantage of second approach is that it needs extra space in the form of `uid2pids` map, though it helps in retaining information (such as OS assigned PIDs). The disadvantage of the first approach is that it fully replaces PIDs with UIDs and hence several changes (though doable) can be expected out of it. Let me know if you have any other ideas or feedbacks on the above solutions. Thanks.

cc: @dharhas @pcmoritz @wuisawesome

**Comment by wuisawesome** [comment]
Can we try to keep the approach a bit simpler? It seems to me like we only need to use this new id while the worker starts up. Can we get away with just doing 

```
    std::unordered_map<StartupToken, StartingWorkerProcessInfo> starting_worker_processes;
```

We can store this token with the rest of the worker information and use it whenever we need to index into that datastructure. That way, I don't think we need to change anything else.

**Comment by czgdp1807** [comment]
We need the UID at two places, one at the time of startup we will pass this `uid` to the worker process via command line arguments. And then we need this `uid` when the worker process sends a registration request (which I think happens somewhere [here](https://github.com/ray-project/ray/blob/4af07a8917723f19a6c20c0d834d99e64b4ddccc/src/ray/raylet_client/raylet_client.cc#L101-L105).

As of now `starting_worker_process` is of type, `std::unordered_map<Process, StartingWorkerProcessInfo>`. If you suggest to change it to `std::unordered_map<uid_t, StartingWorkerProcessInfo>`(assuming by `StartupToken` you meant the new id which we are using) then probably we will be required to make changes wherever `starting_worker_process` is used (due to change in its type).

> We can store this token with the rest of the worker information

If by this you mean to store `uid` along with `pid`  in `ProcessFD` class then indexing would still be difficult because of the difference in `pid` of wrapper process and the actual worker process. Though if only use `uid` for indexing then I think it is possible by defining the `operator()`for `Process` class and defining an `operator==` method inside `Process` class (something like in [this answer](https://stackoverflow.com/a/17017281)). Then `pid` won't be considered (which is the cause of the issue) while indexing `starting_worker_processes`. **Overall, this should not require many changes just some additional methods in `Process` class and sharing of `uid` in Registration request sent by the worker process.** Let me know what you think. Thanks for the comments.

**Comment by wuisawesome** [comment]
I mean storing it in the `Worker` class. I believe (other than before the worker process is registered) you always have access to the full Worker object when using that map. (i.e. right now we always do `worker->GetShimProcess()`, do `worker->GetStartupToken()` instead).

**Comment by wuisawesome** [comment]
(using startup token instead of `uid` because we have a ton of id types in Ray already and uid isn't very descriptive)

**Comment by pcmoritz** [comment]
Agreed with @wuisawesome, the uid should only be the key in the `starting_worker_process` map [as in std::unordered_map<uid_t, StartingWorkerProcessInfo>], and then when the worker registers itself, it sends its own `pid` in the `RegisterWorker` message, which is then used by the Raylet after that (we need the `pid` e.g. for killing processes). I don't think we need additional datastructures here, but please let us know @czgdp1807 if you run into problems with this. I don't think we need to store the token in the `Worker` class, after the worker registers we can use the `pid` and don't need changes, right?

Also calling it `StartupToken` instead of `uid` is preferred.

**Comment by czgdp1807** [comment]
> I don't think we need to store the token in the `Worker` class, after the worker registers we can use the `pid` and don't need changes, right?

Yeah, if we are storing the `StarupToken` a.k.a `uid` as key then I don't think it is needed anywhere else. We just need to check whether it is the worker process which we are trying to register. Since, `StartupToken` will match if that will be the case, so once registered we don't need to worry about it.
I will open a WIP PR soon and then we can see if it works out well or not. Thanks for the feedbacks.
